### PR TITLE
Add ServiceHub implementation of IDocumentTrackingService

### PR DESCRIPTION
**Pull request by sharwell**
This implementation helps identify problematic uses of transient state in remote invocations that are intended to be stateless.

### CODE REVIEW

**Review by CyrusNajmabadi — APPROVED** [review]
great idea :)

File: src/Workspaces/Remote/ServiceHub/Services/ServiceHubDocumentTrackingService.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        public event EventHandler<DocumentId?> ActiveDocumentChanged { add { } remove { } }
+        public event EventHandler<EventArgs> NonRoslynBufferTextChanged { add { } remove { } }
+
+        public ImmutableArray<DocumentId> GetVisibleDocuments()
+        {
+            Fail("Code should not be attempting to obtain visible documents from a stateless remote invocation.");
+            return ImmutableArray<DocumentId>.Empty;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/ServiceHubDocumentTrackingService.cs:36]: why?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttributeDiscovery/RemoteDesignerAttributeDiscoveryService.cs [hunk 2]
```diff
@@ -36,7 +36,7 @@ public ValueTask StartScanningForDesignerAttributesAsync(RemoteServiceCallbackId
                     analyzerProvider,
                     new IncrementalAnalyzerProviderMetadata(
                         nameof(RemoteDesignerAttributeIncrementalAnalyzerProvider),
-                        highPriorityForActiveFile: true,
+                        highPriorityForActiveFile: false,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/DesignerAttributeDiscovery/RemoteDesignerAttributeDiscoveryService.cs:39]: so this is correct, but still problematic :-/   we *do* want a system whereby a user touches a document, and we update the 'designable' state for it in VS asap.  Do we have any thoughts on how we might accomplish that moving forward?

File: src/Workspaces/Remote/ServiceHub/Services/ServiceHubDocumentTrackingService.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/Remote/ServiceHub/Services/ServiceHubDocumentTrackingService.cs:36]: Accidental. I needed to find the name `WatsonReporter.ReportNonFatal`. I wasn't sure what the name was, but I knew it used `MethodImplOptions.NoInlining` as part of setting the stack trace. I just forgot to remove this line after finding references.

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/ServiceHubDocumentTrackingService.cs:36]: :D

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttributeDiscovery/RemoteDesignerAttributeDiscoveryService.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttributeDiscovery/RemoteDesignerAttributeDiscoveryService.cs:39]: Diagnostic incremental analyzer handles this by running the solution crawler portion in **devenv.exe** and delegating the individual document/project analysis operations to OOP. For designer attribute discovery, we have a few options:

1. Rely on Solution Crawler to minimize the number of document analysis operations that are requested following a document change. Analysis operations would directly prioritize open documents, but open documents would scan quickly following changes by nature of work being added to a short queue.
2. Move designer attribute solution crawling to **devenv.exe**, and delegate to OOP similar to diagnostics.
3. Leave this analyzer unchanged (essentially option 1), but add a parallel API that can perform targeted updates of a document following context switches.

**Pull request merged**